tests/test_rio_create.py

```diff
@@ -0,0 +1,354 @@
+        assert dataset.shape == (512, 256)
+        assert dataset.count == 3
+        assert dataset.dtypes == ("uint8", "uint8", "uint8")
+        assert dataset.driver == "GTiff"
+        assert all(((256, 128) == hw for hw in dataset.block_shapes))
+
+
+def test_create_no_overwrite(tmp_path, runner):
+    """Don't allow overwrite of existing file without option."""
+    outpath = tmp_path.joinpath("out.tif")
+    outpath.touch()
+    outfile = str(outpath)
+
+    result = runner.invoke(
+        main_group,
+        [
+            "create",
+            "-f",
+            "GTiff",
+            "-t",
+            "uint8",
+            "-n",
+            "3",
+            "-h",
+            "512",
+            "-w",
+            "256",
+            outfile,
+        ],
+    )
+    assert result.exit_code == 1
+    assert "File exists and won't be overwritten" in result.output
+
+
+def test_create_overwrite(tmp_path, runner):
+    """Allow overwrite of existing file with option."""
+    outpath = tmp_path.joinpath("out.tif")
+    outpath.touch()
+    outfile = str(outpath)
+
+    result = runner.invoke(
+        main_group,
+        [
+            "create",
+            "-f",
+            "GTiff",
+            "-t",
+            "uint8",
+            "-n",
+            "3",
+            "-h",
+            "512",
+            "-w",
+            "256",
+            "--overwrite",
+            outfile,
+        ],
+    )
+    assert result.exit_code == 0
+
+
+def test_create_no_overwrite_nonfile(runner):
+    """Don't allow overwrite of existing non-file without option."""
+    with MemoryFile(bytes(bytearray(100000))) as memfile:
+        result = runner.invoke(
+            main_group,
+            [
+                "create",
+                "-f",
+                "GTiff",
+                "-t",
+                "uint8",
+                "-n",
+                "3",
+                "-h",
+                "512",
+                "-w",
+                "256",
+                memfile.name,
+            ],
+        )
+        assert result.exit_code == 1
+        assert "Object exists and won't be overwritten" in result.output
+
+
+def test_create_overwrite_nonfile(runner):
+    """Allow overwrite of existing non-file with option."""
+    with MemoryFile(bytes(bytearray(100000))) as memfile:
+        result = runner.invoke(
+            main_group,
+            [
+                "create",
+                "-f",
+                "GTiff",
+                "-t",
+                "uint8",
+                "-n",
+                "1",
+                "-h",
+                "16",
+                "-w",
+                "16",
+                "--overwrite",
+                memfile.name,
+            ],
+        )
+        assert result.exit_code == 0
+
+        with rasterio.open(memfile.name) as dataset:
+            assert dataset.count == 1
+            assert dataset.height == 16
+            assert dataset.width == 16
```

Comment:
Until today, I hadn't realized that vsimem files are fixed in size. You can clobber a large vsimem file with a smaller one and GDAL will read it (at least for GeoTIFF), but you can't clobber a small vsimem file with a larger one.

Reply:
> but you can't overwrite a small vsimem file with a larger one.

I don't think that's always true. Might depend on what you do exactly (if you create the file with VSIFileFromMemBuffer with bTakeOwnership=FALSE, then as GDAL doesn't own the data, it can't reallocate it to a larger buffer). But:
```
>>> from osgeo import gdal
>>> gdal.FileFromMemBuffer('/vsimem/foo', 'foo')
>>> gdal.FileFromMemBuffer('/vsimem/bar', 'barbar')
>>> print(gdal.VSIStatL('/vsimem/foo').size)
3
>>> gdal.CopyFile('/vsimem/bar', '/vsimem/foo')
>>> print(gdal.VSIStatL('/vsimem/foo').size)
6
```

Reply:
@rouault thanks, that makes sense. And yes, that is what `MemoryFile(bytes(...))` does.